clients/roscpp/include/ros/timer_manager.h

```diff
@@ -428,11 +428,36 @@ void TimerManager<T, D, E>::setPeriod(int32_t handle, const D& period)
   {
     return;
   }
+  
+  if(reset)
+  {
+    info->next_expected = T::now() + period;
+  }
+  
+  // If called in a callback, last_real has not been updated yet so (now - last_real) > period
+  // In this case, let next_expected be updated only in updateNext
```

Comment:
A problem here might be that it is not clear that the callback from which `setPeriod` has been called is triggered by the same timer or if they are unrelated. Only if they are the same the assumption that `last_real` is not up-to-date holds.